### PR TITLE
Change scoring function in sandbox tutorial to directly interact with FS

### PR DESCRIPTION
**Pull request by anli5005**
Previously in the sandbox tutorial, we asked the model to write to stdout and evaluated it on that, in addition to restricting the creation of files in its system prompt. This conflicted with the `hello-world` task of writing "Hello, world!" to a file named hello.txt, causing the model to effectively crash out in its chain of thought.

This PR fixes that by moving the eval and reward functions to directly use the `modal.Sandbox` API and check for the file. This is a stopgap solution while we iterate on the final design for how sandboxes/sandbox pools will look on the Gym, but for now this should help tighten up correctness on this particular tutorial.

---

<a href="https://modal.devinenterprise.com/review/modal-projects/training-gym/pull/354" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: tutorials/tutorial_generator/rl/001_sandboxes.py [hunk 1]
```diff
@@ -144,25 +145,53 @@ def _dataset_preview_code():
...
+            content = sandbox.filesystem.read_text("/app/hello.txt")
+            score = float(content.strip() == EXPECTED_HELLO)
+            metadata = {"hello_txt": content, "stderr": stderr}
+        except modal.exception.SandboxFilesystemNotFoundError:
+            score = 0.0
+            metadata = {"error": "hello.txt was not created", "stderr": stderr}
+        finally:
+            sandbox.terminate()
+            sandbox.detach()
+        return score, metadata
```
[the remark this answers is not in the document]
Comment (anli5005) [reply tutorials/tutorial_generator/rl/001_sandboxes.py:194]: [I extended the try/except to catch SandboxFilesystemError and SandboxTerminatedError](https://github.com/modal-projects/training-gym/pull/354/changes#diff-74534997c6475a27e2ceb5e3861b5db35ada7b42506c6a1f7aec0eaca5839622R110-R115), but not sure if we want to do more

File: tutorials/tutorial_generator/rl/001_sandboxes.py [hunk 2]
```diff
@@ -144,25 +145,53 @@ def _dataset_preview_code():
...
+            "infinity",
+            app=sandbox_app,
+            image=sandbox_image,
+            workdir="/app",
+            timeout=10,
+            cpu=(0.125, 1.0),
+            memory=(128, 1024),
+        )
+        process = sandbox.exec("python", "-c", code, timeout=3)
+        process.wait()
```
[the remark this answers is not in the document]
Comment (anli5005) [reply tutorials/tutorial_generator/rl/001_sandboxes.py]: I don't anticipate this being a problem for this tutorial, seeing as the solution is literally just writing a file

File: tutorials/tutorial_generator/rl/001_sandboxes.py [hunk 3]
```diff
@@ -95,9 +97,9 @@ def _dataset_intro():
     - `task.toml` — metadata (difficulty, category)
     - `tests/` — verification tests (format varies by task)
 
-    The hello-world task uses pytest-based verification rather than
-    `*.in`/`*.out` file pairs, so we define stdin/stdout test cases
-    inline and pass them to `HarborEval` via the `test_cases` field.
+    The hello-world task asks the agent to create `hello.txt` with
+    `Hello, world!` as its content. Its verifier checks the file inside the
```
[the remark this answers is not in the document]
Comment (anli5005) [reply tutorials/tutorial_generator/rl/001_sandboxes.py]: Reworded this section slightly:

https://github.com/modal-projects/training-gym/pull/354/changes#diff-0943769f699144417f16957780d0f45faf900d16aac843d48cbfb8c28dc1f783R98-R100

File: tutorials/tutorial_generator/rl/001_sandboxes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anli5005) [reply tutorials/tutorial_generator/rl/001_sandboxes.py:194]: dupe

File: tutorials/tutorial_generator/rl/001_sandboxes.py [hunk 4]
```diff
@@ -144,25 +143,59 @@ def _dataset_preview_code():
...
+        sandbox_image = modal.Image.debian_slim(python_version="3.12").run_commands(
+            "mkdir -p /app",
+        )
+        sandbox = modal.Sandbox._experimental_create(
+            "sleep",
+            "infinity",
+            app=sandbox_app,
+            image=sandbox_image,
+            workdir="/app",
+            timeout=10,
```
[the remark this answers is not in the document]
Comment (anli5005) [reply tutorials/tutorial_generator/rl/001_sandboxes.py:171]: this is literally the 3rd time you've said this devin

File: tutorials/tutorial_generator/rl/001_sandboxes.py [hunk 5]
```diff
@@ -144,25 +145,53 @@ def _dataset_preview_code():
...
+            cpu=(0.125, 1.0),
```
[the remark this answers is not in the document]
Comment (anli5005) [reply tutorials/tutorial_generator/rl/001_sandboxes.py]: I thought about re-adding some of that verbiage, but much of it relates to how `score_in_sandbox` works, which we're not using here anymore:

> `score_in_sandbox` enforces `sandbox_cpu`/`sandbox_memory` with a `"limit"` policy by default: rather than reserving that capacity up front, the values become burst ceilings, so Modal bills each sandbox by actual CPU-/RAM-second usage instead of the (usually idle) reservation. Pass `cpu_policy="ignore"` to let rollouts burst above
 the configured values, or `"reserve"` for the legacy fixed-reservation behavior.

I feel like the actual CPU/RAM/timeout parameters are reasonably self-explanatory and so in theory safe to leave out, as people can look them up in the docs if they're curious. Down to reconsider though if you have thoughts @andrewhinh

[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/tutorial_generator/rl/001_sandboxes.py]: i think setting defaults is fine, but why do we need to specify a limit here?

File: tutorials/tutorial_generator/rl/001_sandboxes.py [hunk 6]
```diff
@@ -144,25 +143,59 @@ def _dataset_preview_code():
...
+def _sandbox_scorer():
+    def score_hello_file(code):
+        sandbox_app = modal.App.lookup(
+            "training-gym-hello-world",
+            create_if_missing=True,
+        )
+        sandbox_image = modal.Image.debian_slim(python_version="3.12").run_commands(
+            "mkdir -p /app",
+        )
+        sandbox = modal.Sandbox._experimental_create(
```
Comment (andrewhinh) [inline tutorials/tutorial_generator/rl/001_sandboxes.py:165]: why experimental?

**Review by andrewhinh — CHANGES_REQUESTED** [review]

File: tutorials/tutorial_generator/rl/001_sandboxes.py (hunk 6 above)
Comment (anli5005) [reply tutorials/tutorial_generator/rl/001_sandboxes.py:165]: `_experimental_create` is what was being used in `score_in_sandbox`:

https://github.com/modal-projects/training-gym/blob/78f53542a230fec9a9753e524d91cb549512e514/modal_training_gym/common/eval.py#L583

so I figured we should use it here, especially because [V2 Sandboxes](https://modal.com/docs/guide/sandbox-v2) are cool

**Review by andrewhinh — APPROVED** [review]
